frontend/lib/src/dataframes/arrowFormatUtils.ts

```diff
@@ -144,89 +151,147 @@ interface Interval {
+  const timeInSeconds = convertTimestampToSeconds(
+    timestamp,
+    field?.type?.unit ?? TimeUnit.SECOND
+  )
+  return moment.unix(timeInSeconds).utc().toDate()
+}
+
+/**
+ * Formats a time value based on the unit information in the field.
+ *
+ * @param timestamp The time value to format.
+ * @param field The field containing the unit information.
+ * @returns The formatted time value.
+ */
+function formatTime(timestamp: number | bigint, field?: Field): string {
+  const date = convertTimestampToDate(timestamp, field)
+  return moment(date)
     .utc()
-    .format(timeInSeconds % 1 === 0 ? "HH:mm:ss" : "HH:mm:ss.SSS")
+    .format(date.getMilliseconds() === 0 ? "HH:mm:ss" : "HH:mm:ss.SSS")
 }
 
-function formatDuration(data: number | bigint, field?: Field): string {
+function formatDate(date: number | Date): string {
+  const formatPattern = "YYYY-MM-DD"
+
+  // Date values from arrow are already converted to a date object
+  // or a timestamp in milliseconds even if the field unit might indicate a
+  // different unit.
+  // https://github.com/apache/arrow/blob/9e08c57c0986531879aadf7942998d26a94a5d1b/js/src/visitor/get.ts#L167-L171
+  if (
+    date instanceof Date ||
+    (typeof date === "number" && Number.isFinite(date))
+  ) {
+    return moment.utc(date).format(formatPattern)
+  }
+
+  logWarning(`Unsupported date type: ${date}`)
+  return String(date)
+}
+
+/**
+ * Format datetime value from Arrow to string.
+ */
+function formatDatetime(date: number | Date, field?: Field): string {
+  // Datetime values from arrow are already converted to a date object
+  // or a timestamp in milliseconds even if the field unit might indicate a
+  // different unit.
+  // https://github.com/apache/arrow/blob/9e08c57c0986531879aadf7942998d26a94a5d1b/js/src/visitor/get.ts#L174-L190
+  let datetime = moment.utc(date)
+
+  const timezone = field?.type?.timezone
+  if (timezone) {
+    if (moment.tz.zone(timezone)) {
+      // If timezone is a valid timezone name (e.g., "America/New_York")
+      datetime = datetime.tz(timezone)
+    } else {
+      // If timezone is a UTC offset (e.g., "+0500")
+      datetime = datetime.utcOffset(timezone)
+    }
+
+    return datetime.format("YYYY-MM-DD HH:mm:ssZ")
+  }
+  // Return the timestamp without timezone information
+  return datetime.format("YYYY-MM-DD HH:mm:ss")
+}
+
+/**
+ * Formats a duration value based on the unit information in the field.
+ *
+ * @param duration The duration value to format.
+ * @param field The field containing the unit information.
+ * @returns The formatted duration value.
+ */
+function formatDuration(duration: number | bigint, field?: Field): string {
+  // unit: 0 is seconds, 1 is milliseconds, 2 is microseconds, 3 is nanoseconds.
   return moment
-    .duration(convertToSeconds(data, field?.type?.unit ?? 3), "seconds")
+    .duration(
+      convertTimestampToSeconds(
+        duration,
+        field?.type?.unit ?? TimeUnit.NANOSECOND
```

Comment:
nit: in [this line](https://github.com/streamlit/streamlit/pull/9920/files#diff-c2fbf1134b562311e4dbd05f4ee8646f0851871078484e058421831281b47b9dR215), the default `TimeUnit` is `SECOND`. Should they be the same or what is the reason that we use different default values?

Reply:
The reason is that the default unit for time values in pandas (and arrow?) is `SECOND`, and for duration, its `NANOSECOND`. But I don't really expect this case to actually happen. I believe unit is probably always provided for time / duration columns.

Reply:
Ahh ok, thank you! Would you mind adding a quick comment at the two places stating what you have just wrote here, e.g. `// The default is NANOSECOND because that is the default unit for duration in pandas` and `// The default is SECOND because that is the default unit for time values in pandas`